x/wasm/internal/keeper/querier_test.go

```diff
@@ -0,0 +1,131 @@
+	"io/ioutil"
+	"os"
+	"testing"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmwasm/wasmd/x/wasm/internal/types"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	abci "github.com/tendermint/tendermint/abci/types"
+)
+
+func TestQueryContractState(t *testing.T) {
+	type model struct {
+		Key   string `json:"key"`
+		Value string `json:"val"`
+	}
+
+	tempDir, err := ioutil.TempDir("", "wasm")
+	require.NoError(t, err)
+	defer os.RemoveAll(tempDir)
+	ctx, accKeeper, keeper := CreateTestInput(t, false, tempDir)
+
+	deposit := sdk.NewCoins(sdk.NewInt64Coin("denom", 100000))
+	topUp := sdk.NewCoins(sdk.NewInt64Coin("denom", 5000))
+	creator := createFakeFundedAccount(ctx, accKeeper, deposit.Add(deposit))
+	anyAddr := createFakeFundedAccount(ctx, accKeeper, topUp)
+
+	wasmCode, err := ioutil.ReadFile("./testdata/contract.wasm")
+	require.NoError(t, err)
+
+	contractID, err := keeper.Create(ctx, creator, wasmCode)
+	require.NoError(t, err)
+
+	_, _, bob := keyPubAddr()
+	initMsg := InitMsg{
+		Verifier:    anyAddr.String(),
+		Beneficiary: bob.String(),
+	}
+	initMsgBz, err := json.Marshal(initMsg)
+	require.NoError(t, err)
+
+	addr, err := keeper.Instantiate(ctx, creator, contractID, initMsgBz, deposit)
+	require.NoError(t, err)
+	require.Equal(t, "cosmos18vd8fpwxzck93qlwghaj6arh4p7c5n89uzcee5", addr.String())
+
+	contractModel := []types.Model{
+		{Key: "foo", Value: "bar"},
+		{Key: string([]byte{0x0, 0x1}), Value: string([]byte{0x2, 0x3})},
+	}
+	keeper.setContractState(ctx, addr, contractModel)
+
+	q := NewQuerier(keeper)
+	specs := map[string]struct {
+		srcPath          []string
+		srcReq           abci.RequestQuery
+		expModelLen      int
+		expModelContains []model
+		expErr           sdk.Error
+	}{
+		"query all": {
+			srcPath:     []string{QueryGetContractState, addr.String(), QueryMethodContractStateAll},
+			expModelLen: 3,
+			expModelContains: []model{
+				{Key: "foo", Value: "bar"},
+				{Key: string([]byte{0x0, 0x1}), Value: string([]byte{0x2, 0x3})},
+			},
+		},
+		"query raw key": {
+			srcPath:          []string{QueryGetContractState, addr.String(), QueryMethodContractStateRaw},
+			srcReq:           abci.RequestQuery{Data: []byte("foo")},
+			expModelLen:      1,
+			expModelContains: []model{{Key: "foo", Value: "bar"}},
+		},
+		"query raw binary key": {
+			srcPath:          []string{QueryGetContractState, addr.String(), QueryMethodContractStateRaw},
+			srcReq:           abci.RequestQuery{Data: []byte{0x0, 0x1}},
+			expModelLen:      1,
+			expModelContains: []model{{Key: string([]byte{0x0, 0x1}), Value: string([]byte{0x2, 0x3})}},
+		},
+		"query smart": {
+			srcPath:     []string{QueryGetContractState, addr.String(), QueryMethodContractStateSmart},
+			srcReq:      abci.RequestQuery{Data: []byte(`{"raw":{"key":"config"}}`)},
+			expModelLen: 1,
+			//expModelContains: []model{}, // stopping here as contract internals are not stable
+		},
+		"query unknown raw key": {
```

Comment:
Why doesn't this return an empty slice?

Reply:
This was the simplest implementation but not necessarily the best. I was under the assumption that the KV store allows nil values so this result make sense but it does not.

Reply:
Maybe you are right, but I would rather let len(0) = deleted. Even if not strictly true, it gives contracts a way to mark things as deleted

Reply:
I added a patch. And also returning non nil model slices.